### PR TITLE
[#18459] - Add cursor-based input handling to `token-input`

### LINKED ISSUE #18459: Inability to Input new value before the previous value on the send assets page
introduced in https://github.com/status-im/status-mobile/pull/19750

### Steps:
1. Enter any single number into the ETH textbox (e.g., "2").
2. Move the cursor before the entered data and input another value (e.g., "1").

### Actual result:
The new value (e.g., "1") is appended after the previous value (e.g., resulting in "21"). The new value is not inserted before the previous entry.

https://github.com/status-im/status-mobile/assets/52490791/867c9a51-46af-4ed0-87c6-1fc3f07a1732

### Expected result:
The new value (e.g., "1") should be inputted before the previous value (e.g., should display "12"), allowing the user to insert values in between existing entries.

cc: @vkjr

### LINKED ISSUE #18466: [Android] Wrong cursor placement after last value removal on the send assets screen
### Steps:
1. Enter any valid number in the input field
2. Remove this value
3. Check cursor

### Actual result:
After removing the last value, the cursor is located at the end, after the "0."

https://github.com/status-im/status-mobile/assets/52490791/39f293d6-0684-4335-849c-d9b3f9297441

### Expected result:
After removing the last value, the cursor is expected to be located at the beginning.
https://www.figma.com/file/HKncH4wnDwZDAhc4AryK8Y/Wallet-for-Mobile?type=design&node-id=921-233460&mode=design&t=z9evxZ4dd3RJ64Rz-0

### ENV:
Nightly [11 Jan 2024]

### PR DESCRIPTION
**Pull request by ulisesmac**
fixes #18459
fixes #18466

### Summary

This PR adds capabilities to delete and type using the cursor position to the token input.

[Screencast from 2024-01-22 23-04-19.webm](https://github.com/status-im/status-mobile/assets/90291778/14d87fb6-5ced-42c9-9433-12728866fa07)

**NOTE: This PR doesn't handle the active cursor selection typing/erasing** 

NOT HANDLED:
image
It just does nothing if we press the keyboard

### Review notes
Properly handling all possibilites and validations for the active cursor selection is an exhaustive task, and it's a little hard to know what do we exactly expect for every case, it's also an error-prone task.

If the handling in this PR is not the expected, I'd suggest opening a separate issue with the interactions we have and the  result we expect (probably we should align with designers). I believe it's better to merge this PR since it adds parts of the total work that needs to get done.

Do we really want to reimplement this logic? inputs already handle this behavior if we use a regular keyboard (not our custom one).

#### Platforms

- Android
- iOS

### Steps to test

- Open Status
- Go to wallet, pick an address with funds, press send, select and address to send and pick any token.
- The Token input now is able to add numbers in cursor's position.

status: ready

### CODE REVIEW

File: src/status_im/contexts/wallet/send/input_amount/view.cljs [hunk 1]
```diff
@@ -105,7 +142,14 @@
                                         (if (> num-value current-limit-amount)
                                           (reset! input-value (str current-limit-amount))
                                           (reset! input-value v))
-                                        (reagent/flush))))]
+                                        (reagent/flush))))
```
Comment (ulisesmac) [inline src/status_im/contexts/wallet/send/input_amount/view.cljs]: @J-Son89 Just wondering if we really need the existing `reagent/flush` calls?

I tried removing them and I didn't find issues, but I didn't test it deeply.

File: src/status_im/contexts/wallet/send/input_amount/view.cljs [hunk 2]
```diff
@@ -77,6 +108,7 @@
         input-value               (reagent/atom "")
         current-limit             (reagent/atom {:amount   limit-crypto
                                                  :currency token-symbol})
+        input-selection           (reagent/atom {:start 0 :end 0})
```
Comment (ulisesmac) [inline src/status_im/contexts/wallet/send/input_amount/view.cljs]: Atom to track the cursor position :eyes:

File: src/status_im/contexts/wallet/send/input_amount/view.cljs [hunk 3]
```diff
@@ -30,28 +30,54 @@
...
+     (swap! input-selection-atom assoc :start start-idx :end end-idx))))
 
 (defn- normalize-input
-  [current v]
-  (cond
-    (and (string/blank? current) (= v dot))
-    (str "0" v)
+  [current v input-selection-atom]
+  (let [{:keys [start end]} @input-selection-atom]
+    (if (= start end)
```
Comment (ulisesmac) [inline src/status_im/contexts/wallet/send/input_amount/view.cljs:59]: `(= start end)`

I'm only handling the cases when the cursor is not selecting text, validating everything seems really exhaustive.

**Comment by OmarBasem** [comment]
@ulisesmac have you tried setting the selection onChangeText using `setNativeProps`, example: https://github.com/status-im/status-mobile/blob/develop/src/status_im/contexts/chat/messenger/composer/handlers.cljs#L133

**Comment by ulisesmac** [comment]
> @ulisesmac have you tried setting the selection onChangeText using `setNativeProps`, example: https://github.com/status-im/status-mobile/blob/develop/src/status_im/contexts/chat/messenger/composer/handlers.cljs#L133

Hi @OmarBasem !
Thanks for the suggestion.

Is there any advantage if we do this?
I think it's the same as if we just pass `:selection`

**Comment by OmarBasem** [comment]
> @ulisesmac have you tried setting the selection onChangeText using `setNativeProps`, example: https://github.com/status-im/status-mobile/blob/develop/src/status_im/contexts/chat/messenger/composer/handlers.cljs#L133

I think it could be a cleaner and shorter solution. I see multiple new methods like `normalize-input` and submethods within it. I have faced similar problems with text inputs before and did not have to use multiple custom methods.
I am thinking if we set the `cursor-position` [on-selection-change](https://github.com/status-im/status-mobile/blob/develop/src/status_im/contexts/chat/messenger/composer/handlers.cljs#L157)  and update the selection start and end inside [on-change-text](https://github.com/status-im/status-mobile/blob/develop/src/status_im/contexts/chat/messenger/composer/handlers.cljs#L157) that could fix the problem.

So I was suggesting to try that solution as I believe it should be more concise (unless I misunderstood something from the original problem)

**Comment by ulisesmac** [comment]
Thanks again for the suggestion @OmarBasem !

Well, this input isn't calling `on-change-text` when we press the keyboard to add numbers because the keyboard is a  react component.

`on-change-text` is called when we update the input using an external keyboard or the OS's keyboard (but I noticed it works really bad atm).

> I am thinking if we set the cursor-position [on-selection-change](https://github.com/status-im/status-mobile/blob/develop/src/status_im/contexts/chat/messenger/composer/handlers.cljs#L157)

This is what it's being done in this PR, in `on-selection-change` we store the cursor position in an atom.

> and update the selection start and end inside [on-change-text](https://github.com/status-im/status-mobile/blob/develop/src/status_im/contexts/chat/messenger/composer/handlers.cljs#L157) that could fix the problem.

Since (as I said before) `on-change-text` doesn't work, it's being done in `handle-keyboard-press`.

> I  see multiple new methods like normalize-input and submethods within it. 

I didn't add the function `normalize-input`  but created the function `move-input-cursor`, so that when we type a new char, we move the cursor to the expected position, otherwise the cursor will stay in the same position.

I think the logic of this input has to be completely reworked, because when it was implemented it didn't consider the cursor's position for validations, new chars added, and chars erased. I see this PR as a step in that direction.

**Comment by OmarBasem** [comment]
> I think the logic of this input has to be completely reworked, because when it was implemented it didn't consider the cursor's position for validations, new chars added, and chars erased. I see this PR as a step in that direction.

Yeah makes sense. Thanks for the clarification!

**Review by OmarBasem — APPROVED** [review]

**Comment by VolodLytvynenko** [comment]
Hi @ulisesmac , thank you for the PR. I will create separate follow-ups for this PR when I find the expected behavior from the design team. The PR is ready to be merged

**Pull request merged**